File: apps/minecraft/bingo-test/src/main/java/crabcraft/net/bingotest/PlayerProgress.java
package crabcraft.net.bingotest;

import crabcraft.net.crabUtilities.bingo.BingoTask;
import java.util.Collections;
import java.util.EnumSet;
import java.util.List;
import java.util.Objects;
import java.util.Set;

/** Mutable, memory-only progress for one weekly-card test player. */
public final class PlayerProgress {
    private final Set<BingoTask> allowedTasks;
    private final EnumSet<BingoTask> completedTasks = EnumSet.noneOf(BingoTask.class);

    public PlayerProgress(List<BingoTask> allowedTasks) {
        Objects.requireNonNull(allowedTasks, "allowedTasks");
        this.allowedTasks = Set.copyOf(allowedTasks);
        if (allowedTasks.size() != 16 || this.allowedTasks.size() != 16) {
            throw new IllegalArgumentException("Bingo progress requires 16 unique tasks");
        }
    }

    /** Returns true only for the first accepted completion of an allowed task. */
    public boolean complete(BingoTask task) {
        Objects.requireNonNull(task, "task");
        return allowedTasks.contains(task) && completedTasks.add(task);
    }

    public boolean isComplete(BingoTask task) {
        Objects.requireNonNull(task, "task");
        return allowedTasks.contains(task) && completedTasks.contains(task);
    }

    public int completedCount() {
        return completedTasks.size();
    }

    public boolean isChecklistComplete() {
        return completedTasks.size() == allowedTasks.size();
    }

    public Set<BingoTask> completedTasks() {
        return Collections.unmodifiableSet(EnumSet.copyOf(completedTasks));
    }

    public void reset() {
        completedTasks.clear();
    }
}

File: apps/minecraft/bingo-test/src/main/kotlin/crabcraft/net/bingotest/PlayerProgress.kt
package crabcraft.net.bingotest

import crabcraft.net.crabUtilities.bingo.BingoTask
import java.util.Collections
import java.util.EnumSet

/** Mutable, memory-only progress for one weekly-card test player. */
class PlayerProgress(allowedTasks: List<BingoTask>) {
    private val allowedTasks = java.util.Set.copyOf(allowedTasks)
    private val completedTasks = EnumSet.noneOf(BingoTask::class.java)

    init {
        require(allowedTasks.size == 16 && this.allowedTasks.size == 16) {
            "Bingo progress requires 16 unique tasks"
        }
    }

    /** Returns true only for the first accepted completion of an allowed task. */
    fun complete(task: BingoTask): Boolean = task in allowedTasks && completedTasks.add(task)
    fun isComplete(task: BingoTask): Boolean = task in allowedTasks && task in completedTasks
    fun completedCount(): Int = completedTasks.size
    fun isChecklistComplete(): Boolean = completedTasks.size == allowedTasks.size
    fun completedTasks(): MutableSet<BingoTask> = Collections.unmodifiableSet(EnumSet.copyOf(completedTasks))
    fun reset() { completedTasks.clear() }
}
